File: src/main/java/social/alone/server/oauth2/user/UserPrincipalAdapter.java
package social.alone.server.oauth2.user;

import social.alone.server.user.User;
import social.alone.server.user.UserRole;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.oauth2.core.user.OAuth2User;

import java.util.*;
import java.util.stream.Collectors;

public class UserPrincipalAdapter implements OAuth2User, UserDetails {
  private Long id;
  private String email;
  private String password;
  private User user;
  private Collection<? extends GrantedAuthority> authorities;
  private Map<String, Object> attributes;

  public UserPrincipalAdapter(User user) {

    this.id = user.getId();
    this.email = user.getEmail();
    this.password = user.getPassword();
    this.authorities = authorities(user.getRoles());
    this.user = user;
  }

  public static UserPrincipalAdapter create(User user) {
    return new UserPrincipalAdapter(user);
  }

  public static UserPrincipalAdapter create(User user, Map<String, Object> attributes) {
    UserPrincipalAdapter userPrincipalAdapter = UserPrincipalAdapter.create(user);
    userPrincipalAdapter.setAttributes(attributes);
    return userPrincipalAdapter;
  }

  private static Collection<? extends GrantedAuthority> authorities(Set<UserRole> roles) {
    return roles.stream()
            .map(r -> new SimpleGrantedAuthority("ROLE_" + r.name()))
            .collect(Collectors.toSet());
  }

  public Long getId() {
    return id;
  }

  public String getEmail() {
    return email;
  }

  @Override
  public String getPassword() {
    return password;
  }

  @Override
  public String getUsername() {
    return email;
  }

  @Override
  public boolean isAccountNonExpired() {
    return true;
  }

  @Override
  public boolean isAccountNonLocked() {
    return true;
  }

  @Override
  public boolean isCredentialsNonExpired() {
    return true;
  }

  @Override
  public boolean isEnabled() {
    return true;
  }

  @Override
  public Collection<? extends GrantedAuthority> getAuthorities() {
    return authorities;
  }

  @Override
  public Map<String, Object> getAttributes() {
    return attributes;
  }

  public void setAttributes(Map<String, Object> attributes) {
    this.attributes = attributes;
  }

  @Override
  public String getName() {
    return String.valueOf(id);
  }

  public User getUser(){
    return user;
  }
}

File: src/main/kotlin/social/alone/server/auth/oauth2/user/UserPrincipalAdapter.kt
package social.alone.server.auth.oauth2.user

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.oauth2.core.user.OAuth2User
import social.alone.server.user.domain.User
import social.alone.server.user.domain.UserRole

class UserPrincipalAdapter(val user: User) : OAuth2User, UserDetails {
    val id: String = user.id
    val email: String? = user.email
    private val password: String? = user.password
    private val authorities: Collection<GrantedAuthority> = authorities(user.roles)
    private var attributes: Map<String, Any>? = null

    override fun getPassword(): String? {
        return password
    }

    override fun getUsername(): String? {
        return email
    }

    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return true
    }

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return authorities
    }

    override fun getAttributes(): Map<String, Any>? {
        return attributes
    }

    fun setAttributes(attributes: Map<String, Any>) {
        this.attributes = attributes
    }

    override fun getName(): String {
        return id.toString()
    }

    companion object {

        fun create(user: User): UserPrincipalAdapter {
            return UserPrincipalAdapter(user)
        }

        fun create(user: User, attributes: Map<String, Any>): UserPrincipalAdapter {
            val userPrincipalAdapter = UserPrincipalAdapter.create(user)
            userPrincipalAdapter.setAttributes(attributes)
            return userPrincipalAdapter
        }

        private fun authorities(roles: Set<UserRole>): Collection<GrantedAuthority> {
            return roles.map { r -> SimpleGrantedAuthority("ROLE_" + r.name) }
        }
    }
}
